File: libs/go/development_triangle.go
package stonecharts

import (
	"fmt"
	"strings"
)

const (
	dtCellW   = 72.0
	dtCellH   = 28.0
	dtHeaderW = 68.0
	dtHeaderH = 24.0

	dtLightR, dtLightG, dtLightB = 227, 242, 253
	dtDarkR, dtDarkG, dtDarkB    = 21, 101, 192
)

func dtScaleColor(t float64) string {
	r := int(float64(dtLightR) + t*float64(dtDarkR-dtLightR) + 0.5)
	g := int(float64(dtLightG) + t*float64(dtDarkG-dtLightG) + 0.5)
	b := int(float64(dtLightB) + t*float64(dtDarkB-dtLightB) + 0.5)
	return fmt.Sprintf("#%02x%02x%02x", r, g, b)
}

func renderDevelopmentTriangleSVG(spec *ChartSpec) string {
	W, H := int(spec.Width), int(spec.Height)
	theme := spec.theme
	if theme == nil {
		t := lightTheme()
		theme = &t
	}

	tri := spec.Triangle
	var origins []string
	var periods []int
	var values [][]float64
	view := "cumulative"
	valueType := "incurred"
	unit := ""
	if tri != nil {
		origins = tri.Origins
		periods = tri.Periods
		values = tri.Values
		if tri.View != "" {
			view = tri.View
		}
		if tri.ValueType != "" {
			valueType = tri.ValueType
		}
		unit = tri.Unit
	}
	nRows := len(origins)
	nCols := len(periods)

	// --- a11y summary (WP7: include view, valueType, unit) ---
	a11yAttr := ""
	a11yDescStr := ""
	if spec.a11yOn() {
		parts := []string{}
		if spec.Title != "" {
			parts = append(parts, spec.Title+".")
		}
		if unit != "" {
			parts = append(parts, "Unit: "+unit+".")
		}
		parts = append(parts, fmt.Sprintf("Development triangle, view: %s, value type: %s.", view, valueType))
		parts = append(parts, fmt.Sprintf("%d origins, %d periods.", nRows, nCols))
		sum := esc(strings.Join(parts, " "))
		a11yAttr = fmt.Sprintf(` role="img" aria-label="%s"`, sum)
		a11yDescStr = fmt.Sprintf("<desc>%s</desc>", sum)
	}

	mTop := 20.0
	if spec.Title != "" {
		mTop += 26
	}
	if spec.Subtitle != "" {
		mTop += 18
	}
	if unit != "" {
		mTop += 16
	}
	mLeft := 22.0
	mRight := 22.0
	mBottom := 20.0
	if spec.Layout != nil && spec.Layout.Margin != nil {
		m := spec.Layout.Margin
		if m.Top != nil {
			mTop = *m.Top
		}
		if m.Left != nil {
			mLeft = *m.Left
		}
		if m.Right != nil {
			mRight = *m.Right
		}
		if m.Bottom != nil {
			mBottom = *m.Bottom
		}
	}
	_ = mRight
	_ = mBottom

	showFactors := spec.Factors != nil && spec.Factors.Show
	useColor := spec.ColorScaleCfg != nil
	diagOn := spec.Diagonal != nil && spec.Diagonal.Highlight

	var allVals []float64
	for _, row := range values {
		allVals = append(allVals, row...)
	}
	minVal := 0.0
	maxVal := 0.0
	if len(allVals) > 0 {
		minVal = allVals[0]
		maxVal = allVals[0]
		for _, v := range allVals[1:] {
			if v < minVal {
				minVal = v
			}
			if v > maxVal {
				maxVal = v
			}
		}
	}
	valRange := maxVal - minVal
	if valRange <= 0 {
		valRange = 1.0
	}

	gridX := mLeft + dtHeaderW
	gridY := mTop + dtHeaderH

	cellColor := theme.GridColor
	if cellColor == "" {
		cellColor = "#d8d8e0"
	}
	textColor := theme.TitleColor
	if textColor == "" {
		textColor = "#22223a"
	}
	headerColor := theme.AxisLabelColor
	if headerColor == "" {
		headerColor = "#666"
	}
	diagColor := "#e65100"
	annColor := "#d32f2f"

	var p strings.Builder

	font := `font-family="Segoe UI, Helvetica, Arial, sans-serif"`
	if spec.Responsive {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %d %d" preserveAspectRatio="xMidYMid meet" width="100%%" %s>`,
			a11yAttr, W, H, font))
	} else {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d" %s>`,
			a11yAttr, W, H, W, H, font))
	}

	if a11yDescStr != "" {
		p.WriteString(a11yDescStr)
	}

	if theme.Background != "" {
		p.WriteString(fmt.Sprintf(
			`<rect class="sc-bg" x="0" y="0" width="%d" height="%d" fill="%s"/>`,
			W, H, theme.Background))
	}

	ty := 26
	if spec.Title != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-title" x="%s" y="%d" text-anchor="middle" font-size="17" font-weight="600" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.TitleColor, esc(spec.Title)))
		ty += 20
	}
	if spec.Subtitle != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-subtitle" x="%s" y="%d" text-anchor="middle" font-size="12" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.SubtitleColor, esc(spec.Subtitle)))
		ty += 16
	}

	// --- WP7: unit label ---
	if unit != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-dt-unit" x="%s" y="%d" text-anchor="middle" font-size="11" fill="%s">Unit: %s</text>`,
			f1(float64(W)/2), ty, theme.SubtitleColor, esc(unit)))
	}

	// --- WP7: wrapping group with data attributes ---
	p.WriteString(fmt.Sprintf(
		`<g class="sc-dt-triangle" data-triangle-view="%s" data-triangle-value-type="%s">`,
		esc(view), esc(valueType)))

	// Period headers
	p.WriteString(`<g class="sc-dt-headers">`)
	for c := 0; c < nCols; c++ {
		cx := gridX + float64(c)*dtCellW + dtCellW/2
		cy := gridY - 6
		p.WriteString(fmt.Sprintf(
			`<text class="sc-dt-period-header" x="%s" y="%s" text-anchor="middle" font-size="11" font-weight="600" fill="%s">%s</text>`,
			f1(cx), f1(cy), headerColor, esc(fmtNum(float64(periods[c])))))
	}
	p.WriteString("</g>")

	// Origin headers
	p.WriteString(`<g class="sc-dt-origins">`)
	for r := 0; r < nRows; r++ {
		ox := gridX - 8
		oy := gridY + float64(r)*dtCellH + dtCellH/2 + 4
		p.WriteString(fmt.Sprintf(
			`<text class="sc-dt-origin-header" x="%s" y="%s" text-anchor="end" font-size="11" font-weight="600" fill="%s">%s</text>`,
			f1(ox), f1(oy), headerColor, esc(origins[r])))
	}
	p.WriteString("</g>")

	// Grid cells
	p.WriteString(`<g class="sc-dt-grid">`)
	for r := 0; r < nRows; r++ {
		var row []float64
		if r < len(values) {
			row = values[r]
		}
		p.WriteString(fmt.Sprintf(`<g class="sc-dt-row" data-origin="%d">`, r))
		for c := 0; c < len(row); c++ {
			v := row[c]
			x := gridX + float64(c)*dtCellW
			y := gridY + float64(r)*dtCellH
			fill := "none"
			if useColor {
				t := (v - minVal) / valRange
				fill = dtScaleColor(t)
			}
			p.WriteString(fmt.Sprintf(
				`<rect class="sc-dt-cell sc-point" data-origin="%d" data-period="%d" x="%s" y="%s" width="%s" height="%s" fill="%s" stroke="%s"/>`,
				r, c, f1(x), f1(y), f1(dtCellW), f1(dtCellH), fill, cellColor))
			tx := x + dtCellW/2
			tyCell := y + dtCellH/2 + 4
			vColor := textColor
			if useColor && (v-minVal)/valRange > 0.55 {
				vColor = "#ffffff"
			}
			p.WriteString(fmt.Sprintf(
				`<text class="sc-dt-value" x="%s" y="%s" text-anchor="middle" font-size="10" fill="%s">%s</text>`,
				f1(tx), f1(tyCell), vColor, esc(fmtNum(v))))
		}
		p.WriteString("</g>")
	}
	p.WriteString("</g>")

	// --- WP5: fixed latest-diagonal (rightmost populated cell) ---
	if diagOn {
		p.WriteString(`<g class="sc-dt-diagonal">`)
		for r := 0; r < nRows; r++ {
			var row []float64
			if r < len(values) {
				row = values[r]
			}
			if len(row) == 0 {
				continue
			}
			c := len(row) - 1
			x := gridX + float64(c)*dtCellW
			y := gridY + float64(r)*dtCellH
			p.WriteString(fmt.Sprintf(
				`<rect class="sc-dt-diag" data-origin="%d" data-period="%d" x="%s" y="%s" width="%s" height="%s" fill="none" stroke="%s" stroke-width="2"/>`,
				r, c, f1(x), f1(y), f1(dtCellW), f1(dtCellH), diagColor))
		}
		if spec.Diagonal != nil && spec.Diagonal.Label != "" {
			lx := gridX + float64(nCols)*dtCellW + 8
			ly := gridY + dtCellH/2 + 4
			p.WriteString(fmt.Sprintf(
				`<text class="sc-dt-diag-label" x="%s" y="%s" font-size="10" fill="%s">%s</text>`,
				f1(lx), f1(ly), diagColor, esc(spec.Diagonal.Label)))
		}
		p.WriteString("</g>")
	}

	// --- WP6: render supplied factors (no computation) ---
	if showFactors && spec.Factors != nil {
		factorValues := spec.Factors.Values
		fy := gridY + float64(nRows)*dtCellH
		flx := gridX - 8
		fly := fy + dtCellH/2 + 4
		p.WriteString(`<g class="sc-dt-factors">`)
		p.WriteString(fmt.Sprintf(
			`<text class="sc-dt-factor-header" x="%s" y="%s" text-anchor="end" font-size="10" font-weight="600" fill="%s">Factors</text>`,
			f1(flx), f1(fly), headerColor))
		for c := 0; c < len(factorValues); c++ {
			fx := gridX + float64(c)*dtCellW + dtCellW/2 + dtCellW/2
			p.WriteString(fmt.Sprintf(
				`<text class="sc-dt-factor" x="%s" y="%s" text-anchor="middle" font-size="10" fill="%s">%s</text>`,
				f1(fx), f1(fly), textColor, fmt.Sprintf("%.3f", factorValues[c])))
		}
		p.WriteString("</g>")
	}

	// --- WP7: annotations with text in accessible metadata ---
	if len(spec.Annotations) > 0 {
		originIdx := make(map[string]int, len(origins))
		for i, o := range origins {
			originIdx[o] = i
		}
		periodIdx := make(map[int]int, nCols)
		for i := 0; i < nCols; i++ {
			periodIdx[periods[i]] = i
		}
		p.WriteString(`<g class="sc-dt-annotations">`)
		for _, ann := range spec.Annotations {
			ri, riOk := originIdx[ann.Origin]
			ci, ciOk := periodIdx[ann.Period]
			if riOk && ciOk {
				var row []float64
				if ri < len(values) {
					row = values[ri]
				}
				if ci < len(row) {
					ax := gridX + float64(ci)*dtCellW + dtCellW - 6
					ay := gridY + float64(ri)*dtCellH + 6
					escapedText := esc(ann.Text)
					p.WriteString(fmt.Sprintf(
						`<g class="sc-dt-annotation-group" aria-label="%s">`,
						escapedText))
					p.WriteString(fmt.Sprintf(`<title>%s</title>`, escapedText))
					p.WriteString(fmt.Sprintf(
						`<circle class="sc-dt-annotation" cx="%s" cy="%s" r="4" fill="%s" opacity="0.8"/>`,
						f1(ax), f1(ay), annColor))
					p.WriteString(fmt.Sprintf(
						`<text class="sc-dt-annotation-text" x="%s" y="%s" text-anchor="middle" font-size="7" font-weight="700" fill="#ffffff">!</text>`,
						f1(ax), f1(ay+3)))
					p.WriteString("</g>")
				}
			}
		}
		p.WriteString("</g>")
	}

	// Close the wrapping triangle group
	p.WriteString("</g>")

	p.WriteString("</svg>")
	return p.String()
}

func devTriangleDataTable(spec *ChartSpec) string {
	tri := spec.Triangle
	caption := ""
	if spec.Title != "" {
		caption = "<caption>" + esc(spec.Title) + "</caption>"
	}
	var head strings.Builder
	for _, per := range tri.Periods {
		head.WriteString(`<th scope="col">` + esc(fmtNum(float64(per))) + "</th>")
	}
	var rows strings.Builder
	for i, origin := range tri.Origins {
		var row []float64
		if i < len(tri.Values) {
			row = tri.Values[i]
		}
		rows.WriteString(`<tr><th scope="row">` + esc(origin) + "</th>")
		for j := 0; j < len(tri.Periods); j++ {
			if j < len(row) {
				rows.WriteString("<td>" + esc(fmtNum(row[j])) + "</td>")
			} else {
				rows.WriteString("<td></td>")
			}
		}
		rows.WriteString("</tr>")
	}
	return `<table class="sc-visually-hidden">` + caption +
		"<thead><tr><td></td>" + head.String() + "</tr></thead>" +
		"<tbody>" + rows.String() + "</tbody></table>"
}

File: libs/python/stonecharts/charts/development_triangle.py
"""Development-triangle chart renderer: ChartSpec -> SVG string.

Actuarial loss-development matrix with optional diagonal highlight,
development factors, color scale, and cell annotations. Does NOT use
render_cartesian — this is a matrix/table layout with its own SVG shell.
See charts/development-triangle/design.md for the full geometry contract.
"""

from __future__ import annotations

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import a11y_summary

CELL_W = 72.0
CELL_H = 28.0
HEADER_W = 68.0
HEADER_H = 24.0

_LIGHT_R, _LIGHT_G, _LIGHT_B = 227, 242, 253
_DARK_R, _DARK_G, _DARK_B = 21, 101, 192


def _scale_color(t: float) -> str:
    r = int(_LIGHT_R + t * (_DARK_R - _LIGHT_R) + 0.5)
    g = int(_LIGHT_G + t * (_DARK_G - _LIGHT_G) + 0.5)
    b = int(_LIGHT_B + t * (_DARK_B - _LIGHT_B) + 0.5)
    return f"#{r:02x}{g:02x}{b:02x}"


def render_svg(spec: ChartSpec) -> str:
    W, H = spec.width, spec.height
    theme = spec.theme
    cid = esc(spec.id)

    a11y_attr = ""
    a11y_desc = ""
    if spec.a11y:
        _sum = esc(a11y_summary(spec, "Development triangle"))
        a11y_attr = f' role="img" aria-label="{_sum}"'
        a11y_desc = f"<desc>{_sum}</desc>"

    m_top: float = 20
    if spec.title:
        m_top += 26
    if spec.subtitle:
        m_top += 18
    m_left: float = 22
    m_right: float = 22
    m_bottom: float = 20
    if spec.layout and spec.layout.margin:
        m = spec.layout.margin
        if m.top is not None:
            m_top = m.top
        if m.left is not None:
            m_left = m.left
        if m.right is not None:
            m_right = m.right
        if m.bottom is not None:
            m_bottom = m.bottom

    tri = spec.triangle
    origins = tri.origins if tri else []
    periods = tri.periods if tri else []
    values = tri.values if tri else []
    n_rows = len(origins)
    n_cols = len(periods)

    show_factors = spec.factors_config is not None and spec.factors_config.show
    use_color = spec.color_scale is not None
    diag_on = spec.diagonal is not None and spec.diagonal.highlight

    all_vals: list[float] = []
    for row in values:
        all_vals.extend(row)
    min_val = min(all_vals) if all_vals else 0.0
    max_val = max(all_vals) if all_vals else 0.0
    val_range = max_val - min_val if max_val > min_val else 1.0

    grid_x = m_left + HEADER_W
    grid_y = m_top + HEADER_H
    factor_h = CELL_H if show_factors else 0.0

    cell_color = theme.grid_color or "#d8d8e0"
    text_color = theme.title_color or "#22223a"
    header_color = theme.axis_label_color or "#666"
    diag_color = "#e65100"
    ann_color = "#d32f2f"

    p: list[str] = []

    _font = 'font-family="Segoe UI, Helvetica, Arial, sans-serif"'
    if spec.responsive:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'viewBox="0 0 {W} {H}" preserveAspectRatio="xMidYMid meet" '
            f'width="100%" {_font}>'
        )
    else:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'width="{W}" height="{H}" viewBox="0 0 {W} {H}" {_font}>'
        )

    if a11y_desc:
        p.append(a11y_desc)

    if theme.background:
        p.append(f'<rect class="sc-bg" x="0" y="0" width="{W}" height="{H}" fill="{theme.background}"/>')

    ty = 26
    if spec.title:
        p.append(
            f'<text class="sc-title" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="17" font-weight="600" fill="{theme.title_color}">{esc(spec.title)}</text>'
        )
        ty += 20
    if spec.subtitle:
        p.append(
            f'<text class="sc-subtitle" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="12" fill="{theme.subtitle_color}">{esc(spec.subtitle)}</text>'
        )

    p.append('<g class="sc-dt-headers">')
    for c in range(n_cols):
        cx = grid_x + c * CELL_W + CELL_W / 2
        cy = grid_y - 6
        p.append(
            f'<text class="sc-dt-period-header" x="{cx:.1f}" y="{cy:.1f}" '
            f'text-anchor="middle" font-size="11" font-weight="600" fill="{header_color}">'
            f'{esc(fmt_num(float(periods[c])))}</text>'
        )
    p.append("</g>")

    p.append('<g class="sc-dt-origins">')
    for r in range(n_rows):
        ox = grid_x - 8
        oy = grid_y + r * CELL_H + CELL_H / 2 + 4
        p.append(
            f'<text class="sc-dt-origin-header" x="{ox:.1f}" y="{oy:.1f}" '
            f'text-anchor="end" font-size="11" font-weight="600" fill="{header_color}">'
            f'{esc(origins[r])}</text>'
        )
    p.append("</g>")

    p.append('<g class="sc-dt-grid">')
    for r in range(n_rows):
        row = values[r] if r < len(values) else []
        p.append(f'<g class="sc-dt-row" data-origin="{r}">')
        for c in range(len(row)):
            v = row[c]
            x = grid_x + c * CELL_W
            y = grid_y + r * CELL_H
            if use_color:
                t = (v - min_val) / val_range
                fill = _scale_color(t)
            else:
                fill = "none"
            p.append(
                f'<rect class="sc-dt-cell sc-point" data-origin="{r}" data-period="{c}" '
                f'x="{x:.1f}" y="{y:.1f}" width="{CELL_W:.1f}" height="{CELL_H:.1f}" '
                f'fill="{fill}" stroke="{cell_color}"/>'
            )
            tx = x + CELL_W / 2
            ty_cell = y + CELL_H / 2 + 4
            v_color = "#ffffff" if use_color and (v - min_val) / val_range > 0.55 else text_color
            p.append(
                f'<text class="sc-dt-value" x="{tx:.1f}" y="{ty_cell:.1f}" '
                f'text-anchor="middle" font-size="10" fill="{v_color}">{esc(fmt_num(v))}</text>'
            )
        p.append("</g>")
    p.append("</g>")

    if diag_on:
        p.append('<g class="sc-dt-diagonal">')
        for r in range(n_rows):
            c = n_rows - 1 - r
            row = values[r] if r < len(values) else []
            if c < len(row):
                x = grid_x + c * CELL_W
                y = grid_y + r * CELL_H
                p.append(
                    f'<rect class="sc-dt-diag" data-origin="{r}" data-period="{c}" '
                    f'x="{x:.1f}" y="{y:.1f}" width="{CELL_W:.1f}" height="{CELL_H:.1f}" '
                    f'fill="none" stroke="{diag_color}" stroke-width="2"/>'
                )
        if spec.diagonal and spec.diagonal.label:
            lx = grid_x + n_cols * CELL_W + 8
            ly = grid_y + CELL_H / 2 + 4
            p.append(
                f'<text class="sc-dt-diag-label" x="{lx:.1f}" y="{ly:.1f}" '
                f'font-size="10" fill="{diag_color}">{esc(spec.diagonal.label)}</text>'
            )
        p.append("</g>")

    if show_factors:
        factors: list[float] = []
        for c in range(n_cols - 1):
            num = 0.0
            den = 0.0
            for r in range(n_rows):
                row = values[r] if r < len(values) else []
                if c + 1 < len(row):
                    num += row[c + 1]
                    den += row[c]
            factors.append(num / den if den > 0 else 0.0)
        fy = grid_y + n_rows * CELL_H
        flx = grid_x - 8
        fly = fy + CELL_H / 2 + 4
        p.append('<g class="sc-dt-factors">')
        p.append(
            f'<text class="sc-dt-factor-header" x="{flx:.1f}" y="{fly:.1f}" '
            f'text-anchor="end" font-size="10" font-weight="600" fill="{header_color}">Factors</text>'
        )
        for c in range(len(factors)):
            fx = grid_x + c * CELL_W + CELL_W / 2 + CELL_W / 2
            p.append(
                f'<text class="sc-dt-factor" x="{fx:.1f}" y="{fly:.1f}" '
                f'text-anchor="middle" font-size="10" fill="{text_color}">{factors[c]:.3f}</text>'
            )
        p.append("</g>")

    if spec.triangle_annotations:
        origin_idx = {o: i for i, o in enumerate(origins)}
        period_idx = {int(periods[i]): i for i in range(n_cols)}
        p.append('<g class="sc-dt-annotations">')
        for ann in spec.triangle_annotations:
            ri = origin_idx.get(ann.origin)
            ci = period_idx.get(ann.period)
            if ri is not None and ci is not None:
                row = values[ri] if ri < len(values) else []
                if ci < len(row):
                    ax = grid_x + ci * CELL_W + CELL_W - 6
                    ay = grid_y + ri * CELL_H + 6
                    p.append(
                        f'<circle class="sc-dt-annotation" cx="{ax:.1f}" cy="{ay:.1f}" '
                        f'r="4" fill="{ann_color}" opacity="0.8"/>'
                    )
                    p.append(
                        f'<text class="sc-dt-annotation-text" x="{ax:.1f}" y="{ay + 3:.1f}" '
                        f'text-anchor="middle" font-size="7" font-weight="700" fill="#ffffff">!</text>'
                    )
        p.append("</g>")

    p.append("</svg>")
    return "".join(p)
